### PR TITLE
sql: implement ON CONFLICT ON CONSTRAINT

### LINKED ISSUE #28161: sql: support INSERT ... ON CONFLICT ON CONSTRAINT
Postgres [supports](https://www.postgresql.org/docs/9.5/static/sql-insert.html) an extension of `INSERT ... ON CONFLICT` where a constraint name is given instead of a set of columns.

We currently don't support this, but we should:
https://github.com/cockroachdb/cockroach/blob/07d85bd1eef62a78d5be2ac9e7574db88c88e35c/pkg/sql/parser/sql.y#L4808

This would have been a solution for #27981 if we supported it.

cc. @knz for triage. Perhaps @emsal1863 would be interested in working on this?

### PR DESCRIPTION
**Pull request by jordanlewis**
Fixes #28161.

This commit adds support for the ON CONFLICT ON CONSTRAINT form of
INSERT ON CONFLICT, which allows users to explicitly specify "arbiter
indexes" for the conflict detection and resolution rather than using the
ON CONFLICT (columns...) form, which infers the arbiter indexes instead.

Release note (sql change): add support for the ON CONFLICT ON CONSTRAINT
form of INSERT ON CONFLICT. This form is added for compatibility with
PostgreSQL. It permits explicitly selecting an arbiter index for INSERT
ON CONFLICT, rather than inferring one using a column list, which is the
default behavior.

### CODE REVIEW

**Comment by jordanlewis** [comment]
@mgartner, I wasn't totally sure what I was doing with regards to UNIQUE WITHOUT INDEX constraints. I'm hoping you can take a look at what I've done here. Same goes for the partial indexes - I believe the correct behavior here (verified in PG) is to disallow selection of partial indexes explicitly, since they're not considered 'constraints' in Postgres.

It opens a bit of a can of worms though - in PG, UNIQUE indexes don't always have to be constraints - whereas in Cockroach, UNIQUE indexes are always both constraints and indexes. I'm a little concerned that we're getting something wrong here - are we accidentally reporting partial unique indexes as "constraints" when they shouldn't be, for example?

Another difference - when we create a partial unique index in Cockroach, it's given the `_key` suffix. In Postgres, it's given the `_idx` suffix, and cannot later be converted into a true "constraint". I'll write some of this stuff up later if you think it's useful, just wanted to jot down the findings for now here, though.

**Comment by ajwerner** [comment]
> It opens a bit of a can of worms though - in PG, UNIQUE indexes don't always have to be constraints - whereas in Cockroach, UNIQUE indexes are always both constraints and indexes. I'm a little concerned that we're getting something wrong here - are we accidentally reporting partial unique indexes as "constraints" when they shouldn't be, for example?

Indeed this stuff is pretty deeply busted in cockroach. I hacked on this a bit in https://github.com/cockroachdb/cockroach/pull/65825 somewhat towards the end of the last cycle but never got it over the edge. I recall there being some tough questions on breaking backwards compat but it's not currently paged in.

**Review by mgartner — CHANGES_REQUESTED** [review]
Ya, this stuff is a bit wonky because we don't currently distinguish between a unique index and a unique constraint internally like Postgres does. As @ajwerner points out, that's a non-trivial thing to fix, and who knows when we will have the time to address it.

However, I don't think that blocks this PR. Postgres, like CRDB, does not allow a unique constraint and an index to share the same name. So there will never be ambiguity in an `ON CONFLICT ON CONSTRAINT` statement. Once we get around to fixing our handling of unique index and unique constraints, we can prevent users from choosing unique indexes as arbiter constraints in this new syntax.

> are we accidentally reporting partial unique indexes as "constraints" when they shouldn't be, for example?

I assume you're referring to the contents of `pg_catalog.pg_constraint`? IIRC I did this to match the behavior of our unique indexes which appear in `pg_constraint`, even though this is inconsistent with Postgres. I think we could remove partial unique indexes from `pg_constraint` without causing issues, without waiting for #65825 to land. I'd defer to @rafiss on whether that's a good idea or not.

> Another difference - when we create a partial unique index in Cockroach, it's given the _key suffix. In Postgres, it's given the _idx suffix, and cannot later be converted into a true "constraint". I'll write some of this stuff up later if you think it's useful, just wanted to jot down the findings for now here, though.

This should be a simple fix. I've created an issue to track: https://github.com/cockroachdb/cockroach/issues/73556

> Reviewed 5 of 5 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-:-MqKRMjEE2UytHfg8cDG:bktcluz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/sql/opt/optbuilder/insert.go, line 1007 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MqKNi5R6br1ofFdDskn:-MqKNi5R6br1ofFdDsko:b-amnof9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b82670d6168b76e53ac499b602787a0dc3ed2eb6/pkg/sql/opt/optbuilder/insert.go#L1007)):*
> ```Go
> // and system columns are ignored.
> func (mb *mutationBuilder) getArbitersFromOnConflict(onConflict *tree.OnConflict) arbiterSet {
> 	if onConflict == nil {
> ```

It would be cleaner to have a single entry point for getting arbiters (ideally `mb.findArbiters`) and handle the three separate logic paths within that function (breaking into helper functions if necessary). Most of the arbiter related code lives in `mutation_builder_arbiter.go`.
___
*[pkg/sql/opt/optbuilder/insert.go, line 1043 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MqKPNU4Ea0XzfJ9ZIcI:-MqKPNU4Ea0XzfJ9ZIcJ:b-1xh8ik) ([raw file](https://github.com/cockroachdb/cockroach/blob/b82670d6168b76e53ac499b602787a0dc3ed2eb6/pkg/sql/opt/optbuilder/insert.go#L1043)):*
> ```Go
> 					panic(
> 						errors.WithHint(pgerror.Newf(pgcode.WrongObjectType, "unique constraint %q for table %q is partial, "+
> 							"so cannot be used as an arbiter index directly",
> ```

This error message is confusing because it can be used as an arbiter index, just not with the `ON CONSTRAINT` syntax.
___
*[pkg/sql/opt/optbuilder/testdata/unique-checks-insert, line 402 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MqKQjQcBjAkHSvuZdVi:-MqKQjQcBjAkHSvuZdVj:bl2586d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b82670d6168b76e53ac499b602787a0dc3ed2eb6/pkg/sql/opt/optbuilder/testdata/unique-checks-insert#L402)):*
> ```
> # On conflict clause references unique without index constraint explicitly.
> build
> INSERT INTO uniq VALUES (1, 2, 3, 4, 5) ON CONFLICT ON CONSTRAINT unique_w DO NOTHING
> ```

nit: add a similar test in `unique-checks-upsert` with `... DO UPDATE SET ...`
___
*[pkg/sql/opt/optbuilder/testdata/unique-checks-insert, line 1010 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MqKQsuUBwVLkz8mHwW2:-MqKQsuUBwVLkz8mHwW3:bl2586d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b82670d6168b76e53ac499b602787a0dc3ed2eb6/pkg/sql/opt/optbuilder/testdata/unique-checks-insert#L1010)):*
> ```
> 
> # Error when trying to select a partial unique without index constraint
> # explicitly, which is not allowed.
> ```

nit: add a similar test in `unique-checks-upsert` with `... DO UPDATE SET ...`
___
*[pkg/sql/opt/optbuilder/testdata/upsert, line 2313 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MqKQBt7CaMc8HRCjCql:-MqKQBt7CaMc8HRCjCqm:bqro3jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b82670d6168b76e53ac499b602787a0dc3ed2eb6/pkg/sql/opt/optbuilder/testdata/upsert#L2313)):*
> ```
> VALUES (1, 2)
> ON CONFLICT ON CONSTRAINT xyz_pkey DO
> UPDATE SET x=5
> ```

nit: I'd sprinkle these tests in the `Test UPDATE SET expressions.` and `Test DO NOTHING` sections, right under a test with the same behavior, but the column-selection syntax.

**Review by jordanlewis — COMMENTED** [review]
Thanks for the commentary here!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-:-MqaAgjYBd1hcAcs3L1o:b-livjim)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/optbuilder/insert.go, line 1007 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MqKNi5R6br1ofFdDskn:-Mqa7TuyBH5V5e5nFefP:bfnrwbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b82670d6168b76e53ac499b602787a0dc3ed2eb6/pkg/sql/opt/optbuilder/insert.go#L1007)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It would be cleaner to have a single entry point for getting arbiters (ideally `mb.findArbiters`) and handle the three separate logic paths within that function (breaking into helper functions if necessary). Most of the arbiter related code lives in `mutation_builder_arbiter.go`.
</blockquote></details>

Done. I renamed this function findArbiters, moved it to the file you named, and renamed the old findArbiters to be called inferArbitersFromConflictOrds.
___
*[pkg/sql/opt/optbuilder/insert.go, line 1043 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MqKPNU4Ea0XzfJ9ZIcI:-Mqa1B130iBVjrBUsUyB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b82670d6168b76e53ac499b602787a0dc3ed2eb6/pkg/sql/opt/optbuilder/insert.go#L1043)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This error message is confusing because it can be used as an arbiter index, just not with the `ON CONSTRAINT` syntax.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/testdata/unique-checks-insert, line 402 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MqKQjQcBjAkHSvuZdVi:-Mqa2gzM6tJAHUoI2l4R:b-8i6skp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b82670d6168b76e53ac499b602787a0dc3ed2eb6/pkg/sql/opt/optbuilder/testdata/unique-checks-insert#L402)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: add a similar test in `unique-checks-upsert` with `... DO UPDATE SET ...`
</blockquote></details>

I'm not sure what you mean here - I don't see a unique-checks-upsert test file, and I don't think `upsert` lets you do `DO UPDATE SET`... right? Sorry, I must be missing something.
___
*[pkg/sql/opt/optbuilder/testdata/unique-checks-insert, line 1010 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MqKQsuUBwVLkz8mHwW2:-Mqa3D9B-YEA5PFmsShm:bdnpd1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b82670d6168b76e53ac499b602787a0dc3ed2eb6/pkg/sql/opt/optbuilder/testdata/unique-checks-insert#L1010)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: add a similar test in `unique-checks-upsert` with `... DO UPDATE SET ...`
</blockquote></details>

Ditto above comment
___
*[pkg/sql/opt/optbuilder/testdata/upsert, line 2313 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MqKQBt7CaMc8HRCjCql:-Mqa3eje1VN9nc2Ngph7:b-me9u6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/b82670d6168b76e53ac499b602787a0dc3ed2eb6/pkg/sql/opt/optbuilder/testdata/upsert#L2313)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: I'd sprinkle these tests in the `Test UPDATE SET expressions.` and `Test DO NOTHING` sections, right under a test with the same behavior, but the column-selection syntax.
</blockquote></details>

I looked at doing this, but in the end I think it's cleaner to have it in a separate section. So I won't take this nit unless you insist! :) 

I didn't think it's necessary to separately test `DO NOTHING` because the code that handles the constraint selection is orthogonal to the code that handles the ACTION differences.

And the variety of testing I wanted to do here was different from the variety that we needed in the DO UPDATE SET cases (subquery sets and DEFAULT sets).

**Review by mgartner — APPROVED** [review]
:lgtm: Very nice! Apologies for my slow response.

> Reviewed 7 of 8 files at r3, 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-:-MsBVj4P0hSNCIcSqK2C:b5gb6nb)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/sql/opt/optbuilder/mutation_builder_arbiter.go, line 33 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MsBV1104HgxWNuVWJdT:-MsBV1104HgxWNuVWJdU:b-y91tnr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e705111f891f7b243cbd1271e7dbb1089a04bcb/pkg/sql/opt/optbuilder/mutation_builder_arbiter.go#L33)):*
> ```Go
> // panics if no arbiters are found.
> //
> // Arbiter constraints ensure that the columns designated by conflictOrds
> ```

Can you reword this to include ON CONSTRAINT here an in the numbered bullets below? We should probably remove references to `conflictOrds` with something more general like "conflict columns" now that `conflictOrds` is no longer an argument.
___
*[pkg/sql/opt/optbuilder/mutation_builder_arbiter.go, line 71 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MsBSWifEgKIe1-9kdKH:-MsBSWifEgKIe1-9kdKI:bfcxkb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e705111f891f7b243cbd1271e7dbb1089a04bcb/pkg/sql/opt/optbuilder/mutation_builder_arbiter.go#L71)):*
> ```Go
> 				}
> 				arbiters.AddIndex(i)
> 				return arbiters
> ```

nit: you could simplify this to `return makeSingleIndexArbiterSet(mb, i)`
___
*[pkg/sql/opt/optbuilder/mutation_builder_arbiter.go, line 81 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MsBSda94pttJ8FiUShm:-MsBSda94pttJ8FiUShn:baq1t3w) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e705111f891f7b243cbd1271e7dbb1089a04bcb/pkg/sql/opt/optbuilder/mutation_builder_arbiter.go#L81)):*
> ```Go
> 				}
> 				arbiters.AddUniqueConstraint(i)
> 				return arbiters
> ```

nit: you could simplify this to `return makeSingleUniqueConstraintArbiterSet(mb, i)`
___
*[pkg/sql/opt/optbuilder/mutation_builder_arbiter.go, line 112 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MsBTGtm7Y8MOip2wM9a:-MsBTGtm7Y8MOip2wM9b:b-b90ol5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e705111f891f7b243cbd1271e7dbb1089a04bcb/pkg/sql/opt/optbuilder/mutation_builder_arbiter.go#L112)):*
> ```Go
> 		onConflict.Constraint, tableName),
> 		"use the ON CONFLICT (columns...) WHERE <predicate> form to select this partial unique constraint as an arbiter",
> 	)
> ```

nit: the linebreaks and indents of this looks off to me. might be more readable as:

```
return errors.WithHint(
    pgerror.Newf(
        pgcode.WrongObjectType,
        "unique constraint %q for table %q is partial, so cannot be used as an arbiter via the ON CONSTRAINT syntax",
        onConflict.Constraint,
        tableName,
    ),
    "use the ON CONFLICT (columns...) WHERE <predicate> form to select this partial unique constraint as an arbiter",
)
```
___
*[pkg/sql/opt/optbuilder/testdata/unique-checks-insert, line 402 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MqKQjQcBjAkHSvuZdVi:-MsBQAL44SkQ3nBbinui:b-xnsan) ([raw file](https://github.com/cockroachdb/cockroach/blob/b82670d6168b76e53ac499b602787a0dc3ed2eb6/pkg/sql/opt/optbuilder/testdata/unique-checks-insert#L402)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I'm not sure what you mean here - I don't see a unique-checks-upsert test file, and I don't think `upsert` lets you do `DO UPDATE SET`... right? Sorry, I must be missing something.
</blockquote></details>

It's confusing. There's a few optimizer test files with "upsert" in the name that include `INSERT ON CONFLICT DO UPDATE` tests. For example: https://github.com/cockroachdb/cockroach/blob/8a9e560fd4ac8489485c87a266e99e4f1a852deb/pkg/sql/opt/optbuilder/testdata/unique-checks-upsert#L385

You could add a test to that file similar to this test but instead of `DO NOTHING`, it uses `DO UPDATE SET ...`. I guess you might not want to based on your other comment about arbiter selection being orthogonal to the conflict action. I'll leave it up to you to decide.
___
*[pkg/sql/opt/optbuilder/testdata/unique-checks-insert, line 1010 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MqKQsuUBwVLkz8mHwW2:-MsBRR6D2aAjrA1c7DHr:b-hnrez9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b82670d6168b76e53ac499b602787a0dc3ed2eb6/pkg/sql/opt/optbuilder/testdata/unique-checks-insert#L1010)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Ditto above comment
</blockquote></details>

See my response.
___
*[pkg/sql/opt/optbuilder/testdata/upsert, line 2313 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73460#-MqKQBt7CaMc8HRCjCql:-MsBR5JmBJ4MRvHq_xN_:b-81nlyp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b82670d6168b76e53ac499b602787a0dc3ed2eb6/pkg/sql/opt/optbuilder/testdata/upsert#L2313)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I looked at doing this, but in the end I think it's cleaner to have it in a separate section. So I won't take this nit unless you insist! :) 

I didn't think it's necessary to separately test `DO NOTHING` because the code that handles the constraint selection is orthogonal to the code that handles the ACTION differences.

And the variety of testing I wanted to do here was different from the variety that we needed in the DO UPDATE SET cases (subquery sets and DEFAULT sets).
</blockquote></details>

SGTM!

**Review by otan — APPROVED** [review]

**Comment by jordanlewis** [comment]
Thanks for the reviews all!

bors r+
